### ISSUE TITLE
Configs: The process cannot access the file because it is being used by another process.

### ISSUE THREAD

**Issue by stepsword** [MinecraftForge/MinecraftForge#7963]
**Minecraft Version:** 1.16.5

**Forge Version:** 36.2.0

**Logs:** {Link(s) to GitHub Gist with full latest.log and/or crash report}

https://gist.github.com/stepsword/0aeb18b5a5ed044194793e3289f18d31
https://gist.github.com/stepsword/0b06ad20e62c2620473443a330dfa506

Note: the bit in another language translates to: **The process cannot access the file because it is being used by another process.** I don't have a crash log with the english translation unfortunately.

**Steps to Reproduce:**

I cannot reproduce it, but I have had 3 different users report it to me - basically they attempt to load a world and it fails. For one user, updating from 36.1.4 fixed it, but I received another crash report today with  36.2.0 as the forge version.

**Description of issue:**

Crash on creating world due to something with creating config files. Also, since this is a new world being created and the config is being generated for the first time, the file can't really have been open anywhere else - so this error is very confusing, cause it doesn't make sense that a newly created file is giving errors about being in use.

**Comment by TheCurle** [comment]
This gives more credence to our thought of the fml.toml corruption being due to a race condition.

Consider this.. being looked into.

**Comment by grannymc** [comment]
2 crash logs from Sky Factory One and 1 from FTB Ocean Block, latest versions as of this date (Ocean Block is latest beta) https://drive.google.com/drive/folders/1OCwibauJoygL5J0hYLWde5Aw-e949r1R

**Comment by BlayTheNinth** [comment]
I don't know if this is of help in this particular case, but it might be related since in terms of race conditions I just came across something odd with the ConfigWatcher.

Basically, I ran into an issue where config files would constantly correct themselves even on compatible changes. When I checked with a breakpoint at

https://github.com/MinecraftForge/MinecraftForge/blob/72598871762a98c566689acddec539fa369ed86c/fmlcore/src/main/java/net/minecraftforge/fml/config/ConfigFileTypeHandler.java#L146

the commented file config object was completely empty (explaining why it felt the need to correct).

However, if I moved the breakpoint up one line (so it ran before `.load()` is called), the config file would proceed to load correctly. This happened even if the breakpoint was just set to log and evaluate without suspending.

See this video for the issue in action (note: [the breakpoint in my test](https://user-images.githubusercontent.com/1933180/147682829-93596df9-7dd6-4f55-999d-522d9699abd3.png) only prints `"loading config at " + Instant.now() + " in " + Thread.currentThread()`, it does not change anything else)
https://user-images.githubusercontent.com/1933180/147669793-c564f546-ab13-4870-84c3-1c10f9e093e9.mp4

(Minecraft 1.18.1, Forge 39.0.9)

**Comment by TheCurle** [comment]
Well, yes, by checking for correctness before the file is loaded, none of the correctness checks even load.

Null values are always incorrect.

What you've done here is disable the verification system to make the errors stop.

We'll need to search a bit more thoroughly to figure out why these values are null after loading.

**Comment by BlayTheNinth** [comment]
No, I haven't actually changed any code, I'm just moving a breakpoint around. The breakpoint only prints out a timestamp + thread it ran on (it didn't actually call any loading code, sorry that wasn't clear). 

So both cases were running the same code, just by putting the breakpoint before the call to "load", enough of a delay is introduced to work around what appears to be a race condition.

**Comment by TheCurle** [comment]
Interesting.
Load is a blocking operation, so any race condition experienced here must be invoked from before this code.

What IDE are you using to set this breakpoint?

**Comment by BlayTheNinth** [comment]
I'm using IntelliJ IDEA 2021.2.4 (Ultimate Edition).

I've done some further testing and I managed to reproduce the issue outside of dev on a new instance with no mods installed (launched via CurseForge App).

Notably, it seems to happen when the file is edited using Notepad++ (v8.1.9.3). I'm running Windows 10 Home (21H1). When editing the file through another editor (e.g. Windows' default Notepad), the issue does not occur.

1. Start an 1.18.1 instance with Forge 39.0.9 (no mods needed)
2. Open forge-client.toml in Notepad++ and change `alwaysSetupTerrainOffThread` to `true`, then save the file
3. Check latest.log / reload the config file - it corrected the config file even though the config was valid

Another noteworthy thing: it seems changes made via Notepad++ trigger two WatchEvents (count = 1) in quick succession, whereas Notepad only results in one WatchEvent (edit: although this could just be the correction happening so probably insignificant).

(editing the config via IntelliJ is a whole different story, resulting in 1-2 watch events with counts between 1-3, causing reloading to sometimes not happen at all as count >= 2 events are skipped completely)

It's starting to look like this is completely unrelated to the original issue reported (but it may have something to do with the fml.toml corruption you mentioned) so I think I'll move this out into a separate issue.

**Comment by diesieben07** [comment]
Sounds like we need a debounce on the watch events.

**Comment by Jiriki79** [comment]
this seems to be still happening even in forge 40.1.73. ive been trying to play all the mods 7 and it would crash before load up and when i followed the steps above to change the forge-client.toml it started loading for me granted very slow even with a ryzen 9 5900x and rtx 2060 super with almost 13 GB ram dedicated out of my 32 GB ram i have

**Comment by Alle-thebug** [comment]
This exact issue is happening to me and I play ATM8 with forge version 43.2.3
[crash-2023-01-11_15.19.20-client.txt](https://github.com/MinecraftForge/MinecraftForge/files/10393335/crash-2023-01-11_15.19.20-client.txt)

**Comment by afuentes1005** [comment]
Hey guys! I had this issue where it would crash right before loading the main menu and the error was that a different toml every time would say that it is "being used by another process". What fixed it for me was modifying the fml.toml file. I changed versionCheck from false to true and I also changed splashscreen from false to true and it worked for me. Not sure if this helps with all issues but maybe itll help some people

**Comment by JustRed23** [comment]
I had a look at the code for correcting files and think i might know what is going on (do keep in mind that i can be wrong)

https://github.com/MinecraftForge/MinecraftForge/blob/547ce9cb7c63d89462f5d3b44495c457ec72dccb/fmlcore/src/main/java/net/minecraftforge/fml/config/ConfigTracker.java#L60

The reader here, creates a file watcher. My guess is that this watcher sees that the configuration is wrong, so it tries to correct it.
https://github.com/MinecraftForge/MinecraftForge/blob/547ce9cb7c63d89462f5d3b44495c457ec72dccb/fmlcore/src/main/java/net/minecraftforge/fml/config/ConfigFileTypeHandler.java#L138
Here it saves the config, BUT it also gets saved here:
https://github.com/MinecraftForge/MinecraftForge/blob/547ce9cb7c63d89462f5d3b44495c457ec72dccb/fmlcore/src/main/java/net/minecraftforge/fml/config/ConfigTracker.java#L63
My guess would be that when it tries to correct the config, it tries to save again and creates the io error.
